File: server/MovieSocialNetwork/ApplicationServices/src/main/kotlin/services/UserService.kt
package services

import dtos.FollowerDto
import models.User
import org.jetbrains.annotations.NotNull
import repository_contracts.UserRepository
import validators.UserValidator
import java.util.*

class UserService(
  @NotNull private val repository: UserRepository,
  private val validator: UserValidator = UserValidator()
) {
  suspend fun createUser(user: User): User {
    validator.validate(user)
    return repository.createUser(user)
  }

  suspend fun updateUser(user: User) {
    validator.validate(user)
    repository.updateUser(user)
  }

  suspend fun followOtherUser(userId: UUID, otherUserId: UUID) {
    repository.addFollower(userId, otherUserId)
  }

  suspend fun getFollowing(userId: UUID): List<FollowerDto> {
    return repository.getFollowing(userId)!!.map { user -> FollowerDto(user.userId, user.username) }
  }

  suspend fun getFollowers(userId: UUID): List<FollowerDto> {
    return repository.getFollowers(userId)!!.map { user -> FollowerDto(user.userId, user.username) }
  }

  suspend fun unfollowOtherUser(userId: UUID, otherUserId: UUID) {
    repository.removeFollower(userId, otherUserId)
  }

  suspend fun deleteUser(userId: UUID) {
    repository.deleteFollowersAndFollowing(userId)
    repository.deleteReviewVotes(userId)
    repository.deleteReviews(userId)
    repository.deleteUser(userId)
  }

  suspend fun doesUserExist(username: String): Boolean {
    return repository.doesUserExist(username)
  }

  suspend fun getUserByExternalId(externalId: String): User {
    return repository.getUserByExternalId(externalId)
  }

  suspend fun getUserByUsername(username: String): User {
    return repository.getUserByUsername(username)
  }
}

File: client/src/app/services/user.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs';
import {DomainUser} from "@models/domain/domain-user";
import {Follower} from "@models/domain/follower-dto";
import {GeneralUserData} from "@models/stats-dtos/general-user-data";

@Injectable({
  providedIn: 'root'
})
export class UserService {

  private userBaseUrl = 'http://localhost:8080/api/users';

  constructor(private http: HttpClient) {
  }

  getUserById(userId: string): Observable<DomainUser> {
    return this.http.get<DomainUser>(`${this.userBaseUrl}/${userId}`);
  }

  getUserByUsername(username: string): Observable<DomainUser> {
    return this.http.get<DomainUser>(`${this.userBaseUrl}/username/${username}`)
  }

  getUserByExternalId(externalId: string): Observable<DomainUser> {
    return this.http.get<DomainUser>(`${this.userBaseUrl}/external/${externalId}`)
  }

  doesUserExist(username: string): Observable<boolean> {
    return this.http.get<boolean>(`${this.userBaseUrl}/exists/${username}`)
  }

  doesUserExistByEmail(email: string): Observable<boolean> {
    return this.http.get<boolean>(`${this.userBaseUrl}/exists/email/${email}`)
  }

  createUser(user: DomainUser): Observable<DomainUser> {
    return this.http.post<DomainUser>(this.userBaseUrl, user);
  }

  updateUser(user: any): Observable<void> {
    return this.http.put<void>(this.userBaseUrl, user);
  }

  deleteUser(userId: string): Observable<any> {
    return this.http.delete(`${this.userBaseUrl}/${userId}`);
  }

  followUser(userId: string, otherUserId: string): Observable<any> {
    return this.http.put(`${this.userBaseUrl}/follow/${userId}/${otherUserId}`,{});
  }

  unfollowUser(userId: string, otherUserId: string): Observable<any> {
    return this.http.put(`${this.userBaseUrl}/unfollow/${userId}/${otherUserId}`,{});
  }

  getFollowers(userId: string): Observable<Follower[]> {
    return this.http.get<Follower[]>(`${this.userBaseUrl}/followers/${userId}`);
  }

  getFollowing(userId: string): Observable<Follower[]> {
    return this.http.get<Follower[]>(`${this.userBaseUrl}/following/${userId}`);
  }

  getGeneralUserData(userId: string): Observable<GeneralUserData> {
    return this.http.get<GeneralUserData>(`${this.userBaseUrl}/general/${userId}`);
  }
}
